returnn/hdf.py

```diff
@@ -346,6 +364,296 @@ def run(self):
+
+    def run(self, index: int):
+        out_hdfs = list(util.chunks(self.out_hdfs, self.concurrent))[index]
+        splits = list(util.chunks(self.splits, self.concurrent))[index]
+        assert len(out_hdfs) == len(splits)
+        _logging.info(f"Writing audio to {out_hdfs} from {splits}.")
+
+        SimpleHDFWriter = get_returnn_simple_hdf_writer(self.returnn_root.get_path())
+        hdf_writers = [SimpleHDFWriter(filename=out_hdf.get_path(), dim=1) for out_hdf in out_hdfs]
+        segment_whitelists = []
+        for split in splits:
+            with uopen(split, "rt") as f:
+                segments_whitelist = {line.strip() for line in f if len(line.strip()) > 0}
+            segment_whitelists.append(segments_whitelist)
+
+        assert len(segment_whitelists) == len(out_hdfs)
+        all_whitelists = set.union(*segment_whitelists)
+
+        c = corpus.Corpus()
+        c.load(self.bliss_corpus.get_path())
+
+        pool = multiprocessing.Pool(self.num_workers)
+        recs = (
+            # We send every segment of the recording grouped together with the audio to
+            # avoid having to read the audio file more than once.
+            #
+            # We only pass down the required audio/segment metadata instead of the whole
+            # `i6_core.lib.corpus.Segment` object to avoid excessive pickling overhead,
+            # stalling the worker processes (if we don't).
+            (rec.audio, segments)
+            for rec in c.all_recordings()
+            if len(
+                segments := [
+                    (full_name, segment.start, segment.end)
+                    for segment in rec.segments
+                    if (full_name := segment.fullname()) in all_whitelists
+                ]
+            )
+            > 0
+        )
+        # use imap instead of imap_unordered to have reproducible results, even though this
+        # causes a slight speed penalty
+        for results in pool.imap(self._process_seq, recs, chunksize=8):
+            for (segment_name, data), (hdf_writer, segments_whitelist) in itertools.product(
+                results, zip(hdf_writers, segment_whitelists)
+            ):
+                if segment_name not in segments_whitelist:
+                    continue
+                _logging.info(f"Writing {segment_name} to {hdf_writer.filename}.")
+                hdf_writer.insert_batch(
+                    inputs=data.reshape(1, -1, 1),
+                    seq_len=[data.shape[0]],
+                    seq_tag=[segment_name],
+                )
+        for hdf_writer in hdf_writers:
+            hdf_writer.close()
+
+    def _process_seq(self, recording: Tuple[str, Sequence[Tuple[str, float, float]]]) -> List[Tuple[str, np.ndarray]]:
+        audio_file, segments = recording
+
+        assert isinstance(self.multi_channel_strategy, (BlissToPcmHDFJob.Mixdown, BlissToPcmHDFJob.PickNth)), (
+            "unknown multi_channel_strategy"
+        )
```

Comment:
```suggestion
        if not isinstance(self.multi_channel_strategy, (BlissToAudioHDFJob.Mixdown, BlissToPcmHDFJob.PickNth)):
            raise NotImplementedError(f"multi_channel_strategy {self.multi_channel_strategy} is not supported by BlissToAudioHDFJob")
```
But better move this into the constructor so that we fail already at graph creation time and not only at runtime

Reply:
There already is an assertion in `__init__`, but I improved the messages for now.